colcon_sanitizer_reports/_sanitizer_section_part.py

```diff
@@ -0,0 +1,135 @@
+    or
+        ==5504==ERROR: LeakSanitizer: detected memory leaks
+
+        Direct leak of 1 byte(s) in 1 object(s) allocated from:
+            <snip relevant stack trace>
+
+        Indirect leak of 1 byte(s) in 1 object(s) allocated from:
+            <snip irrelevant stack trace>
+
+        SUMMARY: AddressSanitizer: 554924 byte(s) leaked in 6385 allocation(s).
+
+    The first example has two section parts. Part one starts with the lock-order-inversion line and
+    includes all the indented lines until the summary line. It includes two relevant and one
+    irrelevant stack traces to report. Part two is the summary line.
+
+    The second example has four section parts. Part one is just the summary line. Parts two starts
+    with the non-indented "Direct leak" line and includes the following stack trace that is relevant
+    to report. Part three starts with the non-indented "Indirect leak" line and includes the
+    following stack trace that is irrelevant to report. The final part is the summary line.
+
+    See _FIND_RELEVANT_STACK_TRACE_BEGIN_REGEXES_BY_ERROR_NAME for stack trace header search
+    patterns that determine which stack traces are relevant. Different error/warning names have
+    different relevant stack traces.
+
+    After initialization, SanitizerSectionPart includes the following data member.
+
+    relevant_stack_traces:
+        Tuple of all stack traces from the section part that are relevant for generating the report.
+    """
+
+    relevant_stack_traces: Tuple[SanitizerSectionPartStackTrace]
+
+    def __init__(self, *, error_name: str, lines: Tuple[str]) -> None:
+        relevant_stack_traces: List[SanitizerSectionPartStackTrace] = []
+        relevant_stack_trace_lines: Optional[List[str]] = None
+        find_relevant_stack_trace_begin_regexes = (
+            _FIND_RELEVANT_STACK_TRACE_BEGIN_REGEXES_BY_ERROR_NAME[error_name]
+        )
+
+        for line in lines:
+            # Check if we're currently gathering lines of a relevant stack trace.
+            if relevant_stack_trace_lines is not None:
+                match = _FIND_STACK_TRACE_LINE_REGEX.match(line)
+                if match is not None:
+                    relevant_stack_trace_lines.append(line)
+                    continue
+
+                relevant_stack_traces.append(
+                    SanitizerSectionPartStackTrace(lines=tuple(relevant_stack_trace_lines))
+                )
+                relevant_stack_trace_lines = None
+                if len(relevant_stack_traces) == len(find_relevant_stack_trace_begin_regexes):
+                    break
+
+            find_relevant_stack_trace_begin_regex = (
+                find_relevant_stack_trace_begin_regexes[len(relevant_stack_traces)]
+            )
+            match = find_relevant_stack_trace_begin_regex.match(line)
+            if match is not None:
+                relevant_stack_trace_lines = []
```

Comment:
I really had to sit down in order to follow what's going on in this loop. Using the match and list states avoids the need for explicit boolean variables but it might be worth adding comments describing the state indicated by each condition.

Here's what I added locally to help myself review (comments not guaranteed to be formatted):

```python
       for line in lines:
            # Check if we're currently gathering lines of a relevant stack trace.
            if relevant_stack_trace_lines is not None: # that is, we're in the middle of parsing a stack trace.
                match = _FIND_STACK_TRACE_LINE_REGEX.match(line)
                if match is not None: # The current `line` is a stacktrace line.
                    relevant_stack_trace_lines.append(line)
                    continue

                # Implicit `else:` of the above case, we are done processing a stacktrace.
                relevant_stack_traces.append(
                    SanitizerSectionPartStackTrace(lines=tuple(relevant_stack_trace_lines))
                )
                # Re-initialize temporary lines storage.
                relevant_stack_trace_lines = None

                if len(relevant_stack_traces) == len(find_relevant_stack_trace_begin_regexes): # We've found all the stacktraces we're looking for.
                    break
            # Pick the next stack trace beginning to find.
            find_relevant_stack_trace_begin_regex = (
                find_relevant_stack_trace_begin_regexes[len(relevant_stack_traces)]
            )
            match = find_relevant_stack_trace_begin_regex.match(line)
            if match is not None: # We found the start of a new stack trace.
                relevant_stack_trace_lines = []
```

I'm not suggesting that these comments be added verbatim. But it might be worth describing the flow here. I see why the `continue` and `break` are used but they do hurt readability. It would be very easy for me to make a mistake contributing to this code in the future.

Reply:
Sorry about that. I'd meant to revise this further but forgot to make a "TODO fix smell". I have a much cleaner solution in the next revision.